node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1454 @@
+
+//! Implements a `CandidateBackingSubsystem`.
+
+#![recursion_limit="256"]
+
+use std::collections::{HashMap, HashSet};
+use std::convert::TryFrom;
+use std::iter::FromIterator;
+use std::pin::Pin;
+use std::time::Duration;
+
+use bitvec::vec::BitVec;
+use log;
+use futures::{
+	select, FutureExt, SinkExt, StreamExt,
+	channel::{oneshot, mpsc},
+	future::{self, Either},
+	task::{Spawn, SpawnError, SpawnExt},
+};
+use futures_timer::Delay;
+use streamunordered::{StreamUnordered, StreamYield};
+
+use primitives::Pair;
+use keystore::KeyStorePtr;
+use polkadot_primitives::{
+	Hash, BlockNumber,
+	parachain::{
+		AbridgedCandidateReceipt, BackedCandidate, Id as ParaId, ValidatorPair, ValidatorId,
+		ValidatorIndex, HeadData, ValidationCode, SigningContext, PoVBlock,
+		CandidateDescriptor,
+	},
+};
+use polkadot_node_primitives::{
+	FromTableMisbehavior, Statement, SignedFullStatement, MisbehaviorReport, ValidationResult,
+};
+use polkadot_subsystem::{
+	FromOverseer, OverseerSignal, Subsystem, SubsystemContext, SpawnedSubsystem,
+};
+use polkadot_subsystem::messages::{
+	AllMessages, CandidateBackingMessage, CandidateSelectionMessage, SchedulerRoster,
+	RuntimeApiMessage, RuntimeApiRequest, CandidateValidationMessage, ValidationFailed,
+	StatementDistributionMessage, NewBackedCandidate, ProvisionerMessage, ProvisionableData,
+	PoVDistributionMessage,
+};
+use statement_table::{
+	generic::AttestedCandidate as TableAttestedCandidate,
+	Table, Context as TableContextTrait, Statement as TableStatement,
+	SignedStatement as TableSignedStatement,
+};
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	NotInValidatorSet,
+	JobNotFound(Hash),
+	InvalidSignature,
+	#[from]
+	ValidationFailed(ValidationFailed),
+	#[from]
+	Oneshot(oneshot::Canceled),
+	#[from]
+	Mpsc(mpsc::SendError),
+	#[from]
+	Spawn(SpawnError),
+}
+
+/// Holds all data needed for candidate backing job operation.
+struct CandidateBackingJob {
+	/// The hash of the relay parent on top of which this job is doing it's work.
+	parent: Hash,
+	/// Inbound message channel receiving part.
+	rx_to: mpsc::Receiver<ToJob>,
+	/// Outbound message channel sending part.
+	tx_from: mpsc::Sender<FromJob>,
+
+	/// The validation codes of the `ParaId`s we are assigned as validators to.
+	_validation_code: ValidationCode,
+	/// `HeadData`s of the parachains that this validator is assigned to.
+	_head_data: HeadData,
+	/// The `ParaId`s assigned to this validator.
+	assignment: ParaId,
+	/// We issued `Valid` statements on about these candidates.
+	issued_validity: HashSet<Hash>,
+	/// `Some(h)` if this job has already issues `Seconded` statemt for some candidate with `h` hash.
+	seconded: Option<Hash>,
+	/// We have already reported misbehaviors for these validators.
+	reported_misbehavior_for: HashSet<ValidatorIndex>,
+
+	table: Table<TableContext>,
+	table_context: TableContext,
+}
+
+const fn group_quorum(n_validators: usize) -> usize {
+	let mut threshold = (n_validators) / 2;
```

Comment:
```suggestion
	let mut threshold = n_validators / 2;
```

Reply:
The whole function could be rewritten as `(n_validators / 2) + 1` in a single line also